src/data/nat/digits.lean

```diff
@@ -236,10 +238,67 @@ begin
   { simp [of_digits, list.sum_cons, ih], }
 end
 
-/-! ### Inequalities
+/-!
+### Properties
+
+This section contains various lemmas of properties relating to `digits` and `of_digits`.
+-/
+
+lemma digits_eq_nil_iff_eq_zero {b n : ℕ} : digits b n = [] ↔ n = 0 :=
+begin
+  split,
+  { intro h,
+    have : of_digits b (digits b n) = of_digits b [], by rw h,
+    convert this,
+    rw of_digits_digits },
+  { rintro rfl,
+    simp }
+end
 
-This section contains various lemmas with inequalities relating to `digits` and `of_digits`.
- -/
+lemma digits_ne_nil_iff_ne_zero {b n : ℕ} : digits b n ≠ [] ↔ n ≠ 0 :=
+not_congr digits_eq_nil_iff_eq_zero
+
+private lemma digits_last_aux {b n : ℕ} (h : 2 ≤ b) (w : 0 < n) :
+  digits b n = ((n % b) :: digits b (n / b)) :=
+begin
+  rcases b with _|_|b,
+  { finish },
+  { norm_num at h },
+  rcases n with _|n,
+  { norm_num at w },
+  simp,
+end
+
+lemma digits_last {b m : ℕ} (h : 2 ≤ b) (hm : 0 < m) (p q) :
+  (digits b m).last p = (digits b (m/b)).last q :=
+by { simp only [digits_last_aux h hm], rw list.last_cons }
+
+private lemma last_digit_ne_zero_aux (b : ℕ) {m : ℕ} (hm : m ≠ 0) (p) :
+  (digits b m).last p ≠ 0 :=
+begin
+  rcases b with _|_|b,
+  { cases m; finish },
+  { cases m, { finish },
+    simp_rw [digits_one, list.last_repeat_succ 1 m],
+    norm_num },
+  revert hm p,
+  apply nat.strong_induction_on m,
+  intros n IH hn p,
+  have hnpos : 0 < n := nat.pos_of_ne_zero hn,
+  by_cases hnb : n < b + 2,
+  { simp_rw [digits_of_lt b.succ.succ n hnpos hnb],
+    exact nat.pos_iff_ne_zero.mp hnpos },
+  { rw digits_last (show 2 ≤ b + 2, from dec_trivial) hnpos,
+    refine IH _ (nat.div_lt_self hnpos dec_trivial) _ _,
+    { rw ←nat.pos_iff_ne_zero,
+      exact nat.div_pos (le_of_not_lt hnb) dec_trivial },
+    { rw [digits_ne_nil_iff_ne_zero, ←nat.pos_iff_ne_zero],
+      exact nat.div_pos (le_of_not_lt hnb) dec_trivial } },
+end
+
+lemma last_digit_ne_zero (b : ℕ) {m : ℕ} (hm : m ≠ 0) :
```

Comment:
For this, I think `m` can be figured out by later hypotheses, but `b` can't be. Right?

Although in the only use of this lemma so far, `b` could be inferred.

Reply:
That is correct. It depends a bit on the lemma whether we make the argument `b` explicit in this case (most of the time, but not always, `b` can be figured out from the statement you're proving at that time).